File: impls/elixir/lib/mix/tasks/stepA_mal.ex
defmodule Mix.Tasks.StepAMal do
  import Mal.Types
  alias Mal.Function

  # for escript execution
  def main(args) do
    run(args)
  end

  def run(args) do
    env = Mal.Env.new()
    Mal.Env.merge(env, Mal.Core.namespace)
    bootstrap(args, env)
    loop(env)
  end

  defp load_file(file_name, env) do
    read_eval_print("""
      (load-file "#{file_name}")
      """, env)
    exit(:normal)
  end

  defp bootstrap(args, env) do
    # *host-language*
    read_eval_print("(def! *host-language* \"Elixir\")", env)

    # not:
    read_eval_print("""
      (def! not
        (fn* (a) (if a false true)))
      """, env)

    # load-file:
    read_eval_print("""
      (def! load-file
        (fn* (f)
          (eval (read-string (str "(do " (slurp f) "\nnil)")))))
      """, env)

    # cond
    read_eval_print("""
      (defmacro! cond
        (fn* (& xs)
          (if (> (count xs) 0)
            (list 'if (first xs)
              (if (> (count xs) 1)
                (nth xs 1)
                (throw \"odd number of forms to cond\"))
              (cons 'cond (rest (rest xs)))))))"
      """, env)

    Mal.Env.set(env, "eval", %Function{value: fn [ast] ->
      eval(ast, env)
    end})

    case args do
      [file_name | rest] ->
        Mal.Env.set(env, "*ARGV*", list(rest))
        load_file(file_name, env)

      [] ->
        Mal.Env.set(env, "*ARGV*", list([]))
        read_eval_print("(println (str \"Mal [\" *host-language* \"]\"))", env)
    end
  end

  defp loop(env) do
    IO.write(:stdio, "user> ")
    IO.read(:stdio, :line)
      |> read_eval_print(env)
      |> IO.puts

    loop(env)
  end

  defp eval_ast({:list, ast, meta}, env) when is_list(ast) do
    {:list, Enum.map(ast, fn elem -> eval(elem, env) end), meta}
  end

  defp eval_ast({:map, ast, meta}, env) do
    map = for {key, value} <- ast, into: %{} do
      {eval(key, env), eval(value, env)}
    end

    {:map, map, meta}
  end

  defp eval_ast({:vector, ast, meta}, env) do
    {:vector, Enum.map(ast, fn elem -> eval(elem, env) end), meta}
  end

  defp eval_ast({:symbol, symbol}, env) do
    case Mal.Env.get(env, symbol) do
      {:ok, value} -> value
      :not_found -> throw({:error, "'#{symbol}' not found"})
    end
  end

  defp eval_ast(ast, _env), do: ast

  defp read(input) do
    Mal.Reader.read_str(input)
  end

  defp eval_bindings([], env), do: env
  defp eval_bindings([{:symbol, key}, binding | tail], env) do
    evaluated = eval(binding, env)
    Mal.Env.set(env, key, evaluated)
    eval_bindings(tail, env)
  end
  defp eval_bindings(_bindings, _env), do: throw({:error, "Unbalanced let* bindings"})

  defp quasiquote({:list, [{:symbol, "unquote"}, arg], _}), do: arg
  defp quasiquote({:list, [{:symbol, "unquote"}|   _], _}), do: throw({:error, "unquote: arg count"})
  defp quasiquote({:list,   xs, _}), do:                         qq_foldr(xs)
  defp quasiquote({:vector, xs, _}), do: list([{:symbol, "vec"}, qq_foldr(xs)])
  defp quasiquote({:symbol, sym}),    do: list([{:symbol, "quote"}, {:symbol, sym}])
  defp quasiquote({:map, ast, meta}), do: list([{:symbol, "quote"}, {:map, ast, meta}])
  defp quasiquote(ast), do: ast

  defp qq_foldr([]),     do: list([])
  defp qq_foldr([x|xs]), do: qq_loop(x, qq_foldr xs)

  defp qq_loop({:list, [{:symbol, "splice-unquote"}, arg], _}, acc), do: list([{:symbol, "concat"}, arg, acc])
  defp qq_loop({:list, [{:symbol, "splice-unquote"}|   _], _},   _), do: throw({:error, "splice-unquote: arg count"})
  defp qq_loop(elt, acc), do: list([{:symbol, "cons"}, quasiquote(elt), acc])

  defp macro_call?({:list, [{:symbol, key} | _tail], _}, env) do
    case Mal.Env.get(env, key) do
      {:ok, %Function{macro: true}} -> true
      _ -> false
    end
  end
  defp macro_call?(_ast, _env), do: false

  defp do_macro_call({:list, [{:symbol, key} | tail], _}, env) do
    {:ok, %Function{value: macro, macro: true}} = Mal.Env.get(env, key)
    macro.(tail)
      |> macroexpand(env)
  end

  defp macroexpand(ast, env) do
    if macro_call?(ast, env) do
      do_macro_call(ast, env)
    else
      ast
    end
  end

  defp eval({:list, [], _} = empty_ast, _env), do: empty_ast
  defp eval({:list, _list, _meta} = ast, env) do
    case macroexpand(ast, env) do
      {:list, list, meta} -> eval_list(list, env, meta)
      result -> eval_ast(result, env)
    end
  end
  defp eval(ast, env), do: eval_ast(ast, env)

  defp eval_list([{:symbol, "macroexpand"}, ast], env, _), do: macroexpand(ast, env)

  defp eval_list([{:symbol, "if"}, condition, if_true | if_false], env, _) do
    result = eval(condition, env)
    if result == nil or result == false do
      case if_false do
        [] -> nil
        [body] -> eval(body, env)
      end
    else
      eval(if_true, env)
    end
  end

  defp eval_list([{:symbol, "do"} | ast], env, _) do
    ast
      |> List.delete_at(-1)
      |> list
      |> eval_ast(env)
    eval(List.last(ast), env)
  end

  defp eval_list([{:symbol, "def!"}, {:symbol, key}, value], env, _) do
    evaluated = eval(value, env)
    Mal.Env.set(env, key, evaluated)
    evaluated
  end

  defp eval_list([{:symbol, "defmacro!"}, {:symbol, key}, function], env, _) do
    macro = %{eval(function, env) | macro: true}
    Mal.Env.set(env, key, macro)
    macro
  end

  defp eval_list([{:symbol, "let*"}, {list_type, bindings, _}, body], env, _)
  when list_type == :list or list_type == :vector do
    let_env = Mal.Env.new(env)
    eval_bindings(bindings, let_env)
    eval(body, let_env)
  end

  defp eval_list([{:symbol, "fn*"}, {list_type, params, _}, body], env, _)
  when list_type == :list or list_type == :vector do
    param_symbols = for {:symbol, symbol} <- params, do: symbol

    closure = fn args ->
      inner = Mal.Env.new(env, param_symbols, args)
      eval(body, inner)
    end

    %Function{value: closure}
  end

  defp eval_list([{:symbol, "quote"}, arg], _env, _), do: arg

  defp eval_list([{:symbol, "quasiquoteexpand"}, ast], _, _) do
    quasiquote(ast)
  end

  defp eval_list([{:symbol, "quasiquote"}, ast], env, _) do
    ast |> quasiquote
      |> eval(env)
  end

  # (try* A (catch* B C))
  defp eval_list([{:symbol, "try*"}, try_form, {:list, catch_list, _meta}], env, _) do
    eval_try(try_form, catch_list, env)
  end
  defp eval_list([{:symbol, "try*"}, try_form], env, _) do
    eval(try_form, env)
  end
  defp eval_list([{:symbol, "try*"}, _try_form, _], _env, _) do
    throw({:error, "try* requires a list as the second parameter"})
  end

  defp eval_list(ast, env, meta) do
    {:list, [func | args], _} = eval_ast({:list, ast, meta}, env)
    func.value.(args)
  end

  defp eval_try(try_form,
  [{:symbol, "catch*"}, {:symbol, exception}, catch_form], env) do
    try do
      eval(try_form, env)
    catch
      {:error, message}->
        catch_env = Mal.Env.new(env)
        Mal.Env.set(catch_env, exception, {:exception, message})
        eval(catch_form, catch_env)
    end
  end
  defp eval_try(_try_form, _catch_list, _env) do
    throw({:error, "catch* requires two arguments"})
  end

  defp print(value) do
    Mal.Printer.print_str(value)
  end

  defp read_eval_print(:eof, _env), do: exit(:normal)
  defp read_eval_print(line, env) do
    read(line)
      |> eval(env)
      |> print
  catch
    {:error, exception} ->
      IO.puts("Error: #{Mal.Printer.print_str(exception)}")
  end
end

File: impls/prolog/stepA_mal.pl
% -*- mode: prolog; -*- select prolog mode in the emacs text editor

:- initialization(main, main).

:- consult([core, env, printer, reader, types, utils]).

% Read

mal_read(Ast) :-
    write("user> "),
    read_line_to_string(current_input, Line),
    (Line = end_of_file -> throw(exit_repl) ; true),
    (rl_add_history(Line) -> true ; true), % fails for duplicate lines
    read_str(Line, Ast).

% Eval non-empty list depending on their first element.
:- discontiguous eval_list/4.

eval_list(Env, 'def!', Args, Res) :- !,
    check(Args = [Key, Form], "def!: expects 2 arguments, got: ~L", [Args]),
    check(atom(Key), "def!: ~F is not a symbol", [Key]),
    eval(Env, Form, Res),
    env_set(Env, Key, Res).

eval_list(Env, 'let*', Args, Res) :- !,
    check(Args = [Binds, Form], "let*: expects 2 arguments, got: ~L", [Args]),
    check(unbox_seq(Binds, Xs), "let*: ~F is not a sequence", [Binds]),
    env(Env, Let_Env),
    check(map_keyvals(let_loop(Let_Env), Xs), "let*: odd length: ~L", [Binds]),
    eval(Let_Env, Form, Res).

let_loop(Env, Key, Form) :- !,
    check(atom(Key), "let*: ~F is not a key", [Key]),
    eval(Env, Form, Value),
    env_set(Env, Key, Value).

eval_list(Env, if, Args, Res) :- !,
    check(if_assign_args(Args, Form, Then, Else),
          "if: expects 2 or 3 arguments, got: ~L", [Args]),
    eval(Env, Form, Test),
    if_select(Test, Then, Else, Selected),
    eval(Env, Selected, Res).

if_assign_args([Form, Then, Else], Form, Then, Else).
if_assign_args([Form, Then],       Form, Then, nil).

if_select(false, _,    Else, Else) :- !.
if_select(nil,   _,    Else, Else) :- !.
if_select(_,     Then, _,    Then).

eval_list(Env, 'fn*', Args, Res) :- !,
    check(Args = [Params, Form], "fn*: expects 2 arguments, got: ~L", [Args]),
    check(unbox_seq(Params, Keys), "fn*: ~F is not a sequence", [Params]),
    check(maplist(atom, Keys), "fn*: ~F should contains symbols", [Params]),
    mal_fn(apply_fn(Keys, Form, Env), Res).

apply_fn(Keys, Form, Env, Args, Res) :-
    env(Env, Apply_Env),
    check(env_bind(Apply_Env, Keys, Args),
          "cannot apply fn*[~L] to [~L]", [Keys, Args]),
    eval(Apply_Env, Form, Res).

eval_list(Env, do, Args, Res) :- !,
    foldl(do_loop(Env), Args, nil, Res).

do_loop(Env, Elt, _Old_Acc, New_Acc) :- eval(Env, Elt, New_Acc).

eval_list(_, quote, Args, Res) :- !,
    check(Args = [Res], "quote: expects 1 argument, got ~L", [Args]).

eval_list(_, quasiquoteexpand, Args, Res) :- !,
    check(Args = [X], "quasiquoteexpand: expects 1 argument, got: ~L", [Args]),
    quasiquote(X, Res).

eval_list(Env, quasiquote, Args, Res) :- !,
    check(Args = [X], "quasiquote: expects 1 argument, got: ~L", [Args]),
    quasiquote(X, Y),
    eval(Env, Y, Res).

quasiquote(List, Res) :-
    list(Xs, List), !,
    (   Xs = [unquote | Args]
    ->  check(Args = [Res], "unquote: expects 1 argument, got: ", [Args])
    ;   list([], Empty),
        foldr(qq_loop, Empty, Xs, Res)).
quasiquote(Vector, Res) :-
    vector(Xs, Vector), !,
    list([], Empty),
    foldr(qq_loop, Empty, Xs, Y),
    list([vec, Y], Res).
quasiquote(nil,   nil).
quasiquote(true,  true).
quasiquote(false, false).
quasiquote(Symbol_Or_Map, Res) :-
    (atom(Symbol_Or_Map) -> true ; is_map(Symbol_Or_Map)), !,
    list([quote, Symbol_Or_Map], Res).
quasiquote(Anything_Else, Anything_Else).

qq_loop(Elt, Acc, Res) :-
    list(['splice-unquote' | Args], Elt), !,
    check(Args = [X], "splice-unquote: expects 1 argument, got:", [Args]),
    list([concat, X, Acc], Res).
qq_loop(Elt, Acc, Res) :-
    quasiquote(Elt, Quasiquoted),
    list([cons, Quasiquoted, Acc], Res).

eval_list(Env, 'try*', Args, Res) :- !,
    (   Args = [Test]
    ->  eval(Env, Test, Res)
    ;   check(Args = [Test, Catch],
              "try*: expects 1 or 2 arguments, got: ~L", [Args]),
        check(list(['catch*', Key, Form], Catch),
              "try*: ~F is not a catch* list", [Catch]),
        check(atom(Key), "catch*: ~F is not a key", [Key]),
        catch(eval(Env, Test, Res), mal_error(Error),
              (env(Env, Try_Env),
               env_set(Try_Env, Key, Error),
               eval(Try_Env, Form, Res)))).

eval_list(Env,  'defmacro!', Args, Res) :- !,
    check(Args = [Key, Form],
          "defmacro!: expects 2 arguments, got: ~L", [Args]),
    check(atom(Key), "defmacro!: ~F is not a key", [Key]),
    eval(Env, Form, Fn),
    check(mal_fn(_Goal, Fn), "defmacro!: ~F is not a function", [Fn]),
    mal_macro(Fn, Res),
    env_set(Env, Key, Res).

eval_list(Env, macroexpand, Args, Res) :- !,
    check(Args = [X], "macroexpand: expects 1 argument, got: ~L", [Args]),
    macroexpand(Env, X, Res).

macroexpand(Env, Ast, Res) :-
    list([Key | Args], Ast),
    env_get(Env, Key, Macro),
    mal_macro(Fn, Macro), !,
    mal_fn(Goal, Fn),
    call(Goal, Args, New_Ast),
    macroexpand(Env, New_Ast, Res).
macroexpand(_, Ast, Ast).

% apply phase

eval_list(Env, First, Rest, Res) :-
    eval(Env, First, Fn),
    (   mal_macro(F, Fn)
    ->  % If the Fn macro refers to F, apply F then evaluate,
        mal_fn(Goal, F),
        call(Goal, Rest, New_Ast),
        eval(Env, New_Ast, Res)
    ;   % else evaluate arguments, apply Fn.
        check(mal_fn(Goal, Fn), "cannot apply, ~F is not a function", [Fn]),
        maplist(eval(Env), Rest, Args),
        call(Goal, Args, Res)).

% The eval function itself.

% Uncomment this to get a trace with environments.
%% eval(Env, Ast, _) :-
%%     format("EVAL: ~F    in ~V\n", [Ast, Env]),
%%     fail.                       % Proceed with normal alternatives.

eval(Env, List, Res) :-
    list([First | Args], List), !,
    eval_list(Env, First, Args, Res).

eval(_,   nil,    nil).
eval(_,   true,   true).
eval(_,   false,  false).
eval(Env, Symbol, Res) :-
    atom(Symbol), !,
    check(env_get(Env, Symbol, Res), "'~F' not found", [Symbol]).

eval(Env, Vector, Res) :-
    vector(Xs, Vector), !,
    maplist(eval(Env), Xs, Ys),
    vector(Ys, Res).

eval(Env, Map, Res) :-
    map_map(eval(Env), Map, Res).

eval(_, Anything_Else, Anything_Else).

% Print

print(Ast) :- format("~F\n", [Ast]).

% REP

rep(Env) :-
    mal_read(Ast),
    eval(Env, Ast, Evaluated),
    print(Evaluated).

% Main program

repl(Env) :-
    catch(rep(Env), mal_error(X), format("Exception: ~F\n", [X])),
    repl(Env).

re(Env, String) :-
    read_str(String, Ast),
    eval(Env, Ast, _).

define_core_function(Env, Symbol, Core_Function) :-
    mal_fn(wrap_failure(Core_Function), Form),
    env_set(Env, Symbol, Form).

core_eval(Env, [Ast], Res) :- eval(Env, Ast, Res).

main(Argv) :-
    getenv("HOME", Home),
    string_concat(Home, "/.mal-history", History),
    (exists_file(History) -> rl_read_history(History) ; true),

    env(Env),
    core_ns(Core_Ns),
    map_keyvals(define_core_function(Env), Core_Ns),
    define_core_function(Env, eval, core_eval(Env)),

    env_set(Env, '*host-language*', "prolog"),

    re(Env, "(def! not (fn* [a] (if a false true)))"),
    re(Env, "(def! load-file (fn* (f) (eval (read-string (str \"(do \" (slurp f) \"\nnil)\")))))"),
    re(Env, "(defmacro! cond (fn* (& xs) (if (> (count xs) 0) (list 'if (first xs) (if (> (count xs) 1) (nth xs 1) (throw \"odd number of forms to cond\")) (cons 'cond (rest (rest xs)))))))"),

    (   maplist(atom_string, Argv, [Script | Args])

    ->  % If Argv starts with a script, set arguments and load it.
        list(Args, Mal_Argv),
        env_set(Env, '*ARGV*', Mal_Argv),

        format(string(Load_Script), "(load-file \"~s\")", [Script]),
        re(Env, Load_Script)

    ;   % else read from standard input.
        list([], Mal_Argv),
        env_set(Env, '*ARGV*', Mal_Argv),

        re(Env, "(println (str \"Mal [\" *host-language* \"]\"))"),
        catch(repl(Env), exit_repl, nl)
    ),

    (rl_write_history(History) -> true ; true).
